File: src/Day15.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;
import java.util.stream.Collectors;

import static javautils.JavaUtils.*;

@SuppressWarnings({"ExplicitToImplicitClassMigration", "RedundantSuppression"})
public class Day15 {
  public enum ObjectType {
    ROBOT, BOX
  }

  public static class GameObject {
    private JPoint2D pos;
    private final ObjectType type;
    private final JPoint2D size;

    public GameObject(JPoint2D pos, ObjectType type, JPoint2D size) {
      this.pos = pos;
      this.type = type;
      this.size = size;
    }

    public JPoint2D getPos() {
      return pos;
    }

    public void setPos(JPoint2D pos) {
      this.pos = pos;
    }

    public ObjectType getType() {
      return type;
    }

    public JPoint2D getSize() {
      return size;
    }

    public GameObject copy() {
      return new GameObject(pos.copy(), type, size.copy());
    }
  }

  public static Pair<JGrid<Character>, List<Character>> init(String input) {
    String[] splitInput = input.split("\n\n");
    JGrid<Character> grid = JGrid.asCharGrid(Arrays.asList(splitInput[0].split("\n")));
    List<Character> moves = splitInput[1].chars().mapToObj(c -> (char) c).toList();
    return new Pair<>(grid, moves);
  }

  public static int score(List<GameObject> objects) {
    return objects.stream()
        .filter(obj -> obj.getType() == ObjectType.BOX)
        .mapToInt(obj -> obj.getPos().y() * 100 + obj.getPos().x())
        .sum();
  }

  public static GameObject findRobot(List<GameObject> objects) {
    return objects.stream()
        .filter(obj -> obj.getType() == ObjectType.ROBOT)
        .findFirst()
        .orElseThrow();
  }

  public static boolean collides(JGrid<Character> grid, JPoint2D newPos, JPoint2D size) {
    for (int x = newPos.x(); x < newPos.x() + size.x(); x++) {
      for (int y = newPos.y(); y < newPos.y() + size.y(); y++) {
        if (grid.get(x, y) != null && grid.get(x, y) == '#') {
          return true;
        }
      }
    }
    return false;
  }

  public static boolean move(JGrid<Character> grid, GameObject gameObject, List<GameObject> gameObjects, JPoint2D direction) {
    if (direction == null) return true;

    JPoint2D newPos = gameObject.pos.plus(direction);

    if (collides(grid, newPos, gameObject.size)) return true;

    List<GameObject> obstacles = gameObjects.stream()
        .filter(other -> other != gameObject)
        .filter(other -> {
          for (int dx = 0; dx < gameObject.getSize().x(); dx++) {
            for (int dy = 0; dy < gameObject.getSize().y(); dy++) {
              JPoint2D checkPos = newPos.plus(new JPoint2D(dx, dy));
              JPoint2D otherPos = other.getPos();
              if (checkPos.x() >= otherPos.x() && checkPos.x() < otherPos.x() + other.getSize().x() &&
                  checkPos.y() >= otherPos.y() && checkPos.y() < otherPos.y() + other.getSize().y()) {
                return true;
              }
            }
          }
          return false;
        })
        .toList();

    boolean isMovementBlocked = obstacles.stream().anyMatch(obstacle -> move(grid, obstacle, gameObjects, direction));
    if (!obstacles.isEmpty() && isMovementBlocked) return true;

    gameObject.setPos(newPos); // Update position
    return false;
  }

  public static Map<JPoint2D, Character> fillWalls(Map<JPoint2D, Character> map, int width) {
    Map<JPoint2D, Character> newMap = new HashMap<>();
    for (Map.Entry<JPoint2D, Character> entry : map.entrySet()) {
      JPoint2D point = entry.getKey();
      char value = entry.getValue();
      for (int i = 0; i < width; i++) {
        newMap.put(new JPoint2D(point.x() * width + i, point.y()), value == '#' ? '#' : '.');
      }
    }
    return newMap;
  }

  public static List<GameObject> createObjectsAndAddToMap(Map<JPoint2D, Character> map, int width, Map<JPoint2D, Character> expandedMap) {
    List<GameObject> resultingObjects = new ArrayList<>();
    for (Map.Entry<JPoint2D, Character> entry : map.entrySet()) {
      JPoint2D point = entry.getKey();
      char value = entry.getValue();
      if (value != '#' && value != '.') {
        JPoint2D pos = new JPoint2D(point.x() * width, point.y());
        GameObject gameObject = new GameObject(
            pos,
            value == '@' ? ObjectType.ROBOT : ObjectType.BOX,
            value == '@' ? new JPoint2D(1, 1) : new JPoint2D(width, 1)
        );
        expandedMap.put(pos, '.');
        resultingObjects.add(gameObject);
      }
    }
    return resultingObjects;
  }

  public static int runSimulation(JGrid<Character> origGrid, List<Character> moves, int width) {
    Map<JPoint2D, Character> gridData = new HashMap<>(origGrid.data());

    Map<JPoint2D, Character> expandedMap = fillWalls(gridData, width);

    List<GameObject> objects = createObjectsAndAddToMap(gridData, width, expandedMap);
    JGrid<Character> grid = new JGrid<>(expandedMap);

    for (Character move : moves) {
      JPoint2D direction = switch (move) {
        case '>' -> RIGHT;
        case '<' -> LEFT;
        case '^' -> UP;
        case 'v' -> DOWN;
        default -> null;
      };
      List<GameObject> snapshot = objects.stream().map(GameObject::copy).toList();
      if (move(grid, findRobot(objects), objects, direction)) {
        objects.clear();
        objects.addAll(snapshot);
      }
    }

    return score(objects);
  }

  public static void main(String[] ignoredArgs) throws IOException {
    String input = Files.readString(Path.of("src/15.txt"));
    Pair<JGrid<Character>, List<Character>> initResult = init(input);

    JGrid<Character> grid = initResult.first();
    List<Character> moves = initResult.second();

    System.out.println("p1: " + runSimulation(grid, moves, 1));
    System.out.println("p2: " + runSimulation(grid, moves, 2));
  }

  public record Pair<A, B>(A first, B second) {
  }
}

File: src/Day15.kt
import java.io.File

enum class ObjectType {
    ROBOT, BOX
}

fun main() {
    data class GameObject(
        var pos: Point2D,
        val type: ObjectType,
        val size: Point2D,
    )

    fun init(input: String): Pair<Grid<Char>, List<Char>> {
        val (gridLit, movesLit) = input.split("\n\n")
        val grid = gridLit.lines().toGrid()
        val moves = movesLit.lines().joinToString("").toList()
        return grid to moves
    }

    val directionMap = mapOf(
        '>' to RIGHT,
        '<' to LEFT,
        '^' to UP,
        'v' to DOWN
    )

    fun score(objects: List<GameObject>): Int {
        return objects
            .filter { it.type == ObjectType.BOX }
            .sumOf { it.pos.y * 100 + it.pos.x }
    }

    fun List<GameObject>.findRobot(): GameObject = first { it.type == ObjectType.ROBOT }
    fun Grid<Char>.collides(newPos: Point2D, size: Point2D): Boolean {
        for (x in newPos.x until newPos.x + size.x) {
            for (y in newPos.y until newPos.y + size.y) {
                if (get(Point2D(x, y)) == '#') return true
            }
        }
        return false
    }

    fun Grid<Char>.move(
        gameObject: GameObject,
        gameObjects: List<GameObject>,
        direction: Point2D?,
    ): Boolean {
        val dirVector = direction ?: return false
        val newPos = gameObject.pos + dirVector

        if (collides(newPos, gameObject.size)) return false

        val obstacles = gameObjects.filter { other ->
            other != gameObject && (0 until gameObject.size.x).any { dx ->
                (0 until gameObject.size.y).any { dy ->
                    val checkPos = newPos + Point2D(dx, dy)
                    val otherPos = other.pos
                    checkPos.x in otherPos.x until otherPos.x + other.size.x && checkPos.y in otherPos.y until otherPos.y + other.size.y
                }
            }
        }

        val isMovementBlocked = obstacles.any { !move(it, gameObjects, direction) }
        if (obstacles.isNotEmpty() && isMovementBlocked) return false
        gameObject.pos = newPos // Update position
        return true
    }

    fun fillWalls(
        map: Map<Point2D, Char>,
        width: Int,
    ): Map<Point2D, Char> = map
        .flatMap { (point, value) ->
            (0 until width)
                .map { i ->
                    Point2D(point.x * width + i, point.y) to if (value == '#') '#' else '.'
                }
        }
        .toMap()

    fun Map<Point2D, Char>.createObjectsAndAddToMap(
        width: Int,
        expandedMap: MutableMap<Point2D, Char>,
    ): List<GameObject> {
        val resultingObjects =
            filter { (_, value) -> value != '#' && value != '.' }
                .map { (point, value) ->
                    GameObject(
                        pos = Point2D(point.x * width, point.y),
                        type = if (value == '@') ObjectType.ROBOT else ObjectType.BOX,
                        size = if (value == '@') Point2D(1, 1) else Point2D(width, 1)
                    ).also {
                        expandedMap[Point2D(point.x * width, point.y)] = '.'
                    }
                }
                .toList()

        return resultingObjects
    }

    fun runSimulation(
        origGrid: Grid<Char>,
        moves: List<Char>,
        width: Int = 1,
    ): Int {
        val gridData = origGrid.data.toMutableMap()

        val expandedMap = HashMap(fillWalls(gridData, width))

        val objects = gridData.createObjectsAndAddToMap(width, expandedMap).toMutableList()
        val grid = Grid(expandedMap.toMap())

        for (direction in moves.map { directionMap[it] }) {
            val snapshot = objects.map { it.copy() }
            if (!grid.move(objects.findRobot(), objects, direction)) {
                objects.clear()
                objects.addAll(snapshot)
            }
        }

        return score(objects)
    }

    val input = File("/home/finkel/work_self/aoc-2025/src/15.txt").readText()
    val (grid, moves) = init(input)

    println("p1: ${runSimulation(grid, moves)}")
    println("p2: ${runSimulation(grid, moves, width = 2)}")
}